src/picom.c

```diff
@@ -142,43 +145,158 @@ static inline struct managed_win *find_win_all(session_t *ps, const xcb_window_t
 	return w;
 }
 
+enum vblank_callback_action
+collect_vblank_interval_statistics(struct vblank_event *e, void *ud) {
+	auto ps = (session_t *)ud;
+	double vblank_interval = NAN;
+	assert(ps->frame_pacing);
+	assert(ps->vblank_scheduler);
+
+	// TODO(yshui): this naive method of estimating vblank interval does not handle
+	//              the variable refresh rate case very well. This includes the case
+	//              of a VRR enabled monitor; or a monitor that's turned off, in which
+	//              case the vblank events might slow down or stop all together.
+	//              I tried using DPMS to detect monitor power state, and stop adding
+	//              samples when the monitor is off, but I had a hard time to get it
+	//              working reliably, there are just too many corner cases.
+
+	// Don't add sample again if we already collected statistics for this vblank
+	if (ps->last_msc < e->msc) {
+		if (ps->last_msc_instant != 0) {
+			auto frame_count = e->msc - ps->last_msc;
+			auto frame_time =
+			    (int)((e->ust - ps->last_msc_instant) / frame_count);
+			if (frame_count == 1) {
+				render_statistics_add_vblank_time_sample(
+				    &ps->render_stats, frame_time);
+				log_trace("Frame count %lu, frame time: %d us, ust: "
+				          "%" PRIu64 "",
+				          frame_count, frame_time, e->ust);
+			} else {
+				log_trace("Frame count %lu, frame time: %d us, msc: "
+				          "%" PRIu64 ", not adding sample.",
+				          frame_count, frame_time, e->ust);
+			}
+		}
+		ps->last_msc_instant = e->ust;
+		ps->last_msc = e->msc;
+	} else if (ps->last_msc > e->msc) {
+		log_warn("PresentCompleteNotify msc is going backwards, last_msc: "
+		         "%" PRIu64 ", current msc: %" PRIu64,
+		         ps->last_msc, e->msc);
```

Comment:
I think if we hit this spot once we are in trouble as `ps->last_msc` stays the same and next collect_vblank_interval_statistics() would land in the same place unconditionally (also see my comment [1868267108](https://github.com/yshui/picom/pull/1156#issuecomment-1868267108))

Reply:
Yes, I would do something similar to this as well.